rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+    // how to use axum with openSSL
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|_| panic!("could not listen on {}", &address));
+
+    pin_mut!(listener);
+
+    let redirector = https_redirect();
+
+    loop {
+        let tower_service = service.clone();
+        let redirector_service = redirector.clone();
+        let tls_acceptor = ssl_acceptor.clone();
+
+        // Wait for a new tcp connection
+        let (tcp_stream, addr) = listener.accept().await.unwrap();
+
+        tokio::spawn(async move {
+            if is_ssl_stream(&tcp_stream).await {
+                // handle HTTPS connection
+                let ssl = Ssl::new(tls_acceptor.context()).unwrap();
+                let mut tls_stream = SslStream::new(ssl, tcp_stream).unwrap();
+                if let Err(err) = SslStream::accept(Pin::new(&mut tls_stream)).await {
+                    tracing::error!("Error during TSL handshake from {}: {}", addr, err);
+                }
+
+                let stream = TokioIo::new(tls_stream);
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        tower_service.clone().call(request)
+                    });
+
+                let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
+                    .serve_connection_with_upgrades(stream, hyper_service)
+                    .await;
+
+                if let Err(err) = ret {
+                    tracing::error!("Error serving connection from {}: {}", addr, err);
+                }
+            } else {
+                // handle HTTP connection
+                let stream = TokioIo::new(tcp_stream);
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        // check if it is local connection or external
+                        // the to_canonical() converts IPv4-mapped IPv6 addresses
+                        // to plain IPv4, then is_loopback() works correctly for the IPv4 connections
+                        if addr.ip().to_canonical().is_loopback() {
+                            // accept plain HTTP on the local connection
+                            tower_service.clone().call(request)
+                        } else {
+                            // redirect external connections to HTTPS
+                            redirector_service.clone().call(request)
+                        }
+                    });
+
+                let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
+                    .serve_connection_with_upgrades(stream, hyper_service)
+                    .await;
+
+                if let Err(err) = ret {
+                    tracing::error!("Error serving connection from {}: {}", addr, err);
+                }
```

Comment:
I see two identical ending of condition. Only difference is what stream and hyper_service contains.

Reply:
It is not trivial: bear in mind that streams have different types. It is doable, sure, but we need to think a little about it.

Reply:
Yes, the problem is that it uses different data types in each branch. I have no idea how solve that, we probably need to use some templates...

Reply:
Yes, most probably generics is the way to go. However, splitting this stuff into smaller functions (one for SSL and the other for non-SSL) would make the thing more readable.